File: attic/blob_pool/js/pool.ts
/**
 * GLYPH Pool References
 *
 * Pool-based deduplication for strings and objects. Mirrors Go's
 * `glyph.Pool` / `glyph.PoolRegistry`. Pool refs use the
 * `^<PoolID>:<Index>` wire format; pool definitions use
 * `@pool.str id=S1 [...]` and `@pool.obj id=O1 [...]`.
 */

import { GValue, PoolRef } from './types';
import { canonicalizeLoose } from './loose';

export enum PoolKind {
  STRING = 'str',
  OBJECT = 'obj',
}

export class ParsePoolError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'ParsePoolError';
  }
}

export function isPoolRefId(ref: string): boolean {
  if (ref.length < 2) return false;
  const first = ref.charCodeAt(0);
  if (first < 65 || first > 90) return false; // A-Z
  let sawDigit = false;
  for (let i = 1; i < ref.length; i++) {
    const c = ref.charCodeAt(i);
    if (c >= 48 && c <= 57) {
      sawDigit = true;
    } else if (c < 65 || c > 90) {
      return false;
    }
  }
  return sawDigit;
}

export function parsePoolRef(input: string): PoolRef {
  if (!input.startsWith('^')) {
    throw new ParsePoolError('pool ref must start with ^');
  }
  const body = input.slice(1);
  const colon = body.indexOf(':');
  if (colon < 0) {
    throw new ParsePoolError('pool ref must contain colon');
  }
  const poolId = body.slice(0, colon);
  if (!isPoolRefId(poolId)) {
    throw new ParsePoolError(`invalid pool ID: ${poolId}`);
  }
  const idxStr = body.slice(colon + 1);
  const index = Number(idxStr);
  if (!Number.isFinite(index) || !Number.isInteger(index) || index < 0) {
    throw new ParsePoolError(`invalid pool index: ${idxStr}`);
  }
  return { poolId, index };
}

export class Pool {
  readonly id: string;
  readonly kind: PoolKind;
  readonly entries: GValue[] = [];

  constructor(id: string, kind: PoolKind) {
    this.id = id;
    this.kind = kind;
  }

  add(value: GValue): number {
    if (this.kind === PoolKind.STRING && value.type !== 'str') {
      throw new ParsePoolError(`pool ${this.id} is a string pool but got ${value.type}`);
    }
    this.entries.push(value);
    return this.entries.length - 1;
  }

  get(index: number): GValue {
    if (index < 0 || index >= this.entries.length) {
      throw new RangeError(`pool ${this.id}[${index}] out of bounds (len=${this.entries.length})`);
    }
    return this.entries[index];
  }

  get length(): number {
    return this.entries.length;
  }
}

export class PoolRegistry {
  private pools = new Map<string, Pool>();

  register(pool: Pool): void {
    this.pools.set(pool.id, pool);
  }

  get(poolId: string): Pool | undefined {
    return this.pools.get(poolId);
  }

  resolve(ref: PoolRef): GValue {
    const pool = this.pools.get(ref.poolId);
    if (pool === undefined) {
      throw new ParsePoolError(`pool not found: ${ref.poolId}`);
    }
    return pool.get(ref.index);
  }

  ids(): string[] {
    return [...this.pools.keys()].sort();
  }
}

export function emitPool(pool: Pool): string {
  const header = `@pool.${pool.kind} id=${pool.id}`;
  const body = pool.entries.map(v => canonicalizeLoose(v)).join(' ');
  return `${header} [${body}]`;
}

/**
 * Split a pool body like `hello world` or `{a=1} {b=2}` into top-level tokens,
 * respecting quotes and bracket/brace nesting.
 */
function tokenizePoolBody(inner: string): string[] {
  const tokens: string[] = [];
  const n = inner.length;
  let i = 0;
  while (i < n) {
    while (i < n && /\s/.test(inner[i])) i++;
    if (i >= n) break;

    const start = i;
    let depth = 0;
    let inStr = false;
    let esc = false;
    while (i < n) {
      const c = inner[i];
      if (esc) {
        esc = false;
      } else if (inStr) {
        if (c === '\\') esc = true;
        else if (c === '"') inStr = false;
      } else {
        if (c === '"') inStr = true;
        else if (c === '[' || c === '{' || c === '(') depth++;
        else if (c === ']' || c === '}' || c === ')') depth--;
        else if (depth === 0 && /\s/.test(c)) break;
      }
      i++;
    }
    tokens.push(inner.slice(start, i));
  }
  return tokens;
}

function parsePoolEntry(token: string): GValue {
  // Lazy import to break circular dependency with parse.ts
  // eslint-disable-next-line @typescript-eslint/no-require-imports
  const { parseScalarValue } = require('./parse') as typeof import('./parse');
  return parseScalarValue(token);
}

export function parsePool(input: string): Pool {
  const s = input.trim();
  let rest: string;
  let kind: PoolKind;
  if (s.startsWith('@pool.str')) {
    kind = PoolKind.STRING;
    rest = s.slice('@pool.str'.length).replace(/^\s+/, '');
  } else if (s.startsWith('@pool.obj')) {
    kind = PoolKind.OBJECT;
    rest = s.slice('@pool.obj'.length).replace(/^\s+/, '');
  } else {
    throw new ParsePoolError('pool must start with @pool.str or @pool.obj');
  }

  if (!rest.startsWith('id=')) {
    throw new ParsePoolError('pool missing id= field');
  }
  rest = rest.slice('id='.length);

  let i = 0;
  while (i < rest.length && !/\s/.test(rest[i]) && rest[i] !== '[') i++;
  const poolId = rest.slice(0, i);
  if (!isPoolRefId(poolId)) {
    throw new ParsePoolError(`invalid pool ID: ${poolId}`);
  }

  const body = rest.slice(i).replace(/^\s+/, '');
  if (!body.startsWith('[') || !body.endsWith(']')) {
    throw new ParsePoolError('pool body must be bracketed list');
  }

  const inner = body.slice(1, -1).trim();
  const pool = new Pool(poolId, kind);
  if (!inner) return pool;

  for (const token of tokenizePoolBody(inner)) {
    const entry = parsePoolEntry(token);
    if (kind === PoolKind.STRING && entry.type !== 'str') {
      throw new ParsePoolError(`@pool.str ${poolId} entry is not a string: ${entry.type}`);
    }
    pool.entries.push(entry);
  }
  return pool;
}

/**
 * Split a document into leading @pool.* definitions and the remaining value.
 */
export function splitDocument(input: string): { pools: string[]; value: string } {
  let text = input.replace(/^\s+/, '');
  const pools: string[] = [];
  while (text.startsWith('@pool.')) {
    let depth = 0;
    let inStr = false;
    let esc = false;
    let i = 0;
    const n = text.length;
    while (i < n) {
      const c = text[i];
      if (esc) {
        esc = false;
      } else if (inStr) {
        if (c === '\\') esc = true;
        else if (c === '"') inStr = false;
      } else {
        if (c === '"') inStr = true;
        else if (c === '[') depth++;
        else if (c === ']') {
          depth--;
          if (depth === 0) {
            i++;
            break;
          }
        }
      }
      i++;
    }
    if (depth !== 0) {
      throw new ParsePoolError('unbalanced brackets in pool definition');
    }
    pools.push(text.slice(0, i));
    text = text.slice(i).replace(/^\s+/, '');
  }
  return { pools, value: text };
}

export interface ParsedDocument {
  registry: PoolRegistry;
  value: GValue;
}

export function parseDocument(input: string): ParsedDocument {
  const { pools, value } = splitDocument(input);
  const registry = new PoolRegistry();
  for (const block of pools) {
    registry.register(parsePool(block));
  }
  if (!value) {
    throw new ParsePoolError('document has no value');
  }
  const parsed = parsePoolEntry(value);
  return { registry, value: parsed };
}

export function resolvePoolRefs(value: GValue, registry: PoolRegistry): GValue {
  switch (value.type) {
    case 'poolRef':
      return registry.resolve(value.asPoolRef()).clone();
    case 'list':
      return GValue.list(...value.asList().map(v => resolvePoolRefs(v, registry)));
    case 'map':
      return GValue.map(
        ...value.asMap().map(e => ({ key: e.key, value: resolvePoolRefs(e.value, registry) }))
      );
    case 'struct': {
      const sv = value.asStruct();
      return GValue.struct(
        sv.typeName,
        ...sv.fields.map(f => ({ key: f.key, value: resolvePoolRefs(f.value, registry) }))
      );
    }
    case 'sum': {
      const sm = value.asSum();
      const inner = sm.value ? resolvePoolRefs(sm.value, registry) : null;
      return GValue.sum(sm.tag, inner);
    }
    default:
      return value.clone();
  }
}

File: attic/blob_pool/py/pool.py
"""
GLYPH Pool References

Pool-based deduplication for strings and objects. Matches Go's
`glyph.Pool`/`glyph.PoolRegistry` semantics. Pool refs use the
`^<PoolID>:<Index>` wire format; pool definitions use
`@pool.str id=S1 [...]` and `@pool.obj id=O1 [...]`.
"""

from __future__ import annotations

import threading
from enum import Enum
from typing import Dict, List, Optional, Tuple

from .types import GType, GValue, PoolRef


class PoolKind(Enum):
    STRING = "str"
    OBJECT = "obj"


def is_pool_ref_id(ref: str) -> bool:
    """Check if a string is a valid pool ID (uppercase letters + at least one digit)."""
    if len(ref) < 2:
        return False
    if not ("A" <= ref[0] <= "Z"):
        return False
    saw_digit = False
    for c in ref[1:]:
        if c.isdigit():
            saw_digit = True
        elif not ("A" <= c <= "Z"):
            return False
    return saw_digit


class ParsePoolError(ValueError):
    """Raised when pool wire format fails to parse."""


def parse_pool_ref(input: str) -> PoolRef:
    """Parse a pool reference from `^S1:0` format."""
    if not input.startswith("^"):
        raise ParsePoolError("pool ref must start with ^")
    body = input[1:]
    colon = body.find(":")
    if colon < 0:
        raise ParsePoolError("pool ref must contain colon")
    pool_id = body[:colon]
    if not is_pool_ref_id(pool_id):
        raise ParsePoolError(f"invalid pool ID: {pool_id}")
    try:
        index = int(body[colon + 1:])
    except ValueError as e:
        raise ParsePoolError(f"invalid pool index: {body[colon + 1:]}") from e
    return PoolRef(pool_id, index)


class Pool:
    """A single named pool of values."""

    def __init__(self, pool_id: str, kind: PoolKind) -> None:
        self.id = pool_id
        self.kind = kind
        self.entries: List[GValue] = []
        self._lock = threading.RLock()

    def add(self, value: GValue) -> int:
        with self._lock:
            if self.kind == PoolKind.STRING and value.type != GType.STR:
                raise ParsePoolError(
                    f"pool {self.id} is a string pool but got {value.type}"
                )
            self.entries.append(value)
            return len(self.entries) - 1

    def get(self, index: int) -> GValue:
        with self._lock:
            if index < 0 or index >= len(self.entries):
                raise IndexError(
                    f"pool {self.id}[{index}] out of bounds (len={len(self.entries)})"
                )
            return self.entries[index]

    def __len__(self) -> int:
        return len(self.entries)


class PoolRegistry:
    """Thread-safe registry of pools keyed by pool ID."""

    def __init__(self) -> None:
        self._pools: Dict[str, Pool] = {}
        self._lock = threading.RLock()

    def register(self, pool: Pool) -> None:
        with self._lock:
            self._pools[pool.id] = pool

    def get(self, pool_id: str) -> Optional[Pool]:
        with self._lock:
            return self._pools.get(pool_id)

    def resolve(self, ref: PoolRef) -> GValue:
        pool = self.get(ref.pool_id)
        if pool is None:
            raise ParsePoolError(f"pool not found: {ref.pool_id}")
        return pool.get(ref.index)

    def ids(self) -> List[str]:
        with self._lock:
            return sorted(self._pools.keys())


def emit_pool(pool: Pool) -> str:
    """Emit canonical pool-definition wire format."""
    from .loose import canonicalize_loose

    header = f"@pool.{pool.kind.value} id={pool.id}"
    body = " ".join(canonicalize_loose(v) for v in pool.entries)
    return f"{header} [{body}]"


def parse_pool(input: str) -> Pool:
    """Parse a `@pool.str` or `@pool.obj` definition."""
    from .parse import parse  # local import avoids circular dep at module load

    s = input.strip()
    if s.startswith("@pool.str"):
        kind = PoolKind.STRING
        rest = s[len("@pool.str"):].lstrip()
    elif s.startswith("@pool.obj"):
        kind = PoolKind.OBJECT
        rest = s[len("@pool.obj"):].lstrip()
    else:
        raise ParsePoolError("pool must start with @pool.str or @pool.obj")

    if not rest.startswith("id="):
        raise ParsePoolError("pool missing id= field")
    rest = rest[len("id="):]

    i = 0
    while i < len(rest) and not rest[i].isspace() and rest[i] != "[":
        i += 1
    pool_id = rest[:i]
    if not is_pool_ref_id(pool_id):
        raise ParsePoolError(f"invalid pool ID: {pool_id}")

    body = rest[i:].lstrip()
    if not body.startswith("[") or not body.endswith("]"):
        raise ParsePoolError("pool body must be bracketed list")

    list_gv = parse(body)
    if list_gv.type != GType.LIST:
        raise ParsePoolError("pool body did not parse as list")

    pool = Pool(pool_id, kind)
    for entry in list_gv.as_list():
        if kind == PoolKind.STRING and entry.type != GType.STR:
            raise ParsePoolError(
                f"@pool.str {pool_id} entry is not a string: {entry.type}"
            )
        pool.entries.append(entry)
    return pool


def split_document(input: str) -> Tuple[List[str], str]:
    """
    Split a document into @pool.* definition lines and the remaining value.

    Pool definitions appear at the top separated from the value by a blank line
    (or newline). Returns (pool_def_blocks, value_text).
    """
    text = input.lstrip()
    pools: List[str] = []
    while text.startswith("@pool."):
        depth = 0
        i = 0
        n = len(text)
        in_str = False
        esc = False
        while i < n:
            c = text[i]
            if esc:
                esc = False
            elif in_str:
                if c == "\\":
                    esc = True
                elif c == '"':
                    in_str = False
            else:
                if c == '"':
                    in_str = True
                elif c == "[":
                    depth += 1
                elif c == "]":
                    depth -= 1
                    if depth == 0:
                        i += 1
                        break
            i += 1
        if depth != 0:
            raise ParsePoolError("unbalanced brackets in pool definition")
        pools.append(text[:i])
        text = text[i:].lstrip()
    return pools, text


def parse_document(input: str) -> Tuple[PoolRegistry, GValue]:
    """Parse a document with optional leading @pool.* definitions."""
    from .parse import parse

    registry = PoolRegistry()
    pool_blocks, value_text = split_document(input)
    for block in pool_blocks:
        registry.register(parse_pool(block))
    if not value_text:
        raise ParsePoolError("document has no value")
    return registry, parse(value_text)


def resolve_pool_refs(value: GValue, registry: PoolRegistry) -> GValue:
    """Recursively replace POOL_REF values with their resolved pool entries."""
    t = value.type
    if t == GType.POOL_REF:
        return registry.resolve(value.as_pool_ref()).clone()
    if t == GType.LIST:
        return GValue.list_(*[resolve_pool_refs(v, registry) for v in value.as_list()])
    if t == GType.MAP:
        from .types import MapEntry
        return GValue.map_(
            *[MapEntry(e.key, resolve_pool_refs(e.value, registry)) for e in value.as_map()]
        )
    if t == GType.STRUCT:
        from .types import MapEntry
        sv = value.as_struct()
        return GValue.struct(
            sv.type_name,
            *[MapEntry(f.key, resolve_pool_refs(f.value, registry)) for f in sv.fields],
        )
    if t == GType.SUM:
        sm = value.as_sum()
        inner = resolve_pool_refs(sm.value, registry) if sm.value is not None else None
        return GValue.sum(sm.tag, inner)
    return value.clone()
